# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/muhammadahannadeembaig/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

Going with **Lane 2 — Refresh / Content Opportunity Scoring**.

Honestly the main reason is that the starter dataset already lines up with this lane trend direction, freshness, position, all of it's right there, so I don't have to wait
weeks before testing something real. Other lanes sounded interesting too (especially the
clustering one) but this one felt like it actually ends in something useful a list someone
can act on, not just charts and observations.

Keeping this as provisional though. Once I get into the bigger warehouse dataset I might
find the label I'm using here is too weak and switch direction slightly.

## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

**The question:** with way too many pages to check by hand, which ones should someone
actually look at first?

**Unit of analysis:** one page (`content_id`).

**What decision this helps with:** right now there's no real system for this a reviewer
would just be guessing or going by memory. This turns it into an actual ranked list based
on evidence instead of gut feeling.

**Who acts on it:** a content reviewer, basically. They'd start from the top of the list
and go down refresh, expand, or skip depending on the reason attached to each page.

**What a wrong call costs:**
- If I flag something that didn't need fixing, that's wasted time and since review time
  is limited, it means some page that actually needed help got skipped instead.
- If I miss a page that's genuinely dropping off, it just keeps quietly losing traffic
  until someone happens to notice, way later than they should have.

**Why not just a checklist:** because "worth refreshing" isn't one clean number — it's a
mix of things (traffic, position, how old the content is, how thin it is) and none of them
alone tell the full story. That's exactly the kind of thing a model is actually useful for.

## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [2]:
import os

# Am cloning my repo
if not os.path.exists("flyrank-ml-internship"):
    !git clone https://github.com/muhammadahannadeembaig/flyrank-ml-internship.git

%cd flyrank-ml-internship

import pandas as pd

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

print("Total rows in starter dataset:", len(df))
print()

# 1. Pages that are old/stale but still getting real visibility
stale_visible = df[(df["days_since_last_update"] >= 180) & (df["impressions_90d"] >= 500)]
print(f"Stale but still visible pages: {len(stale_visible)} "
      f"({len(stale_visible)/len(df)*100:.1f}% of all pages)")

# 2. Pages that are declining even though there's still real search demand
declining_with_demand = df[(df["trend_direction"] == "down") & (df["impressions_90d"] >= 100)]
print(f"Declining pages with real demand: {len(declining_with_demand)} "
      f"({len(declining_with_demand)/len(df)*100:.1f}% of all pages)")

# 3. Pages that are thin content but still pulling in traffic
thin_visible = df[(df["word_count"] > 0) & (df["word_count"] < 1200) & (df["impressions_90d"] >= 250)]
print(f"Thin content that's still visible: {len(thin_visible)} "
      f"({len(thin_visible)/len(df)*100:.1f}% of all pages)")

Cloning into 'flyrank-ml-internship'...
remote: Enumerating objects: 119, done.
remote: Counting objects: 100% (119/119), done.
remote: Compressing objects: 100% (74/74), done.
remote: Total 119 (delta 36), reused 101 (delta 29), pack-reused 0 (from 0)
Receiving objects: 100% (119/119), 1.84 MiB | 15.20 MiB/s, done.
Resolving deltas: 100% (36/36), done.
/content/flyrank-ml-internship
Total rows in starter dataset: 30000

Stale but still visible pages: 17 (0.1% of all pages)
Declining pages with real demand: 13152 (43.8% of all pages)
Thin content that's still visible: 82 (0.3% of all pages)


Ran this on the 30,000-row starter set and the numbers are pretty uneven:

- **13,152 pages (43.8%)** are marked "declining" while still pulling real search demand
  — that's a huge chunk, way more than I expected.
- Only **17 pages (0.1%)** are stale-but-visible, and **82 (0.3%)** are thin-but-visible —
  both way smaller than the declining group.

So the declining-with-demand signal is clearly the strongest one here, at least on this
slice. The other two might matter more once I'm working with the full warehouse data,
which has way more pages and more history to check freshness against. For now this is
enough evidence that Lane 2 has something real to work with.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

**What I can actually say:**
- Which pages look stale, declining, or thin based on real numbers not guesses.
- A ranked list of what's worth checking first, and the rule behind why each page got
  flagged.
- Whether this scoring beats just doing it randomly or by a simple rule once I actually
  test that later.

**What I can't say:**
- That fixing a flagged page will actually bring traffic back. That needs a real before/after
  test, which I don't have. This is just pointing reviewers in the right direction, not
  proving cause and effect.
- That "declining" the way I've defined it here (`trend_direction == "down"`) is the real
  definition — it's based on the current window, not on what actually happens next. A
  better version later would predict a future window instead of just labeling the present.
- Anything about how Google actually ranks things — I only see outcomes, not the reasoning
  behind them.

## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.